File: digimon_gym/engine/data/scripts/ex4/ex4_073.py
from __future__ import annotations
from typing import TYPE_CHECKING, List, Dict, Any
from ....core.card_script import CardScript
from ....interfaces.card_effect import ICardEffect
from ....data.enums import EffectTiming

if TYPE_CHECKING:
    from ....core.card_source import CardSource


class EX4_073(CardScript):
    """EX4-073 Omnimon Alter-B | Lv.7 Black Digimon | 15000 DP | Cost 15

    Alt digivolve: Lv.7 [Omnimon] for cost 2.

    [When Digivolving] <De-Digivolve 3> 1 of your opponent's Digimon. Then,
        delete up to 6 play cost's total worth of their Digimon.

    [When Attacking] By trashing up to 3 level 6 or higher cards in this
        Digimon's digivolution cards, for each card trashed, activate the
        effect below. Then, if you trashed 3 cards, trash the top 2 cards
        of your opponent's security stack.
        - Delete 1 of your opponent's Digimon or Tamers with the lowest
          play cost.
    """

    def get_card_effects(self, card: 'CardSource') -> List['ICardEffect']:
        effects = []

        def _get_play_cost(perm) -> int:
            if perm.top_card:
                cost = getattr(perm.top_card, 'get_cost_itself', 0)
                return int(cost) if cost else 0
            return 0

        # --- Effect 0: Alt digivolve from Lv.7 Omnimon for cost 2 ---
        effect0 = ICardEffect()
        effect0.set_effect_name("EX4-073 Alternate digivolution requirement")
        effect0.set_effect_description("Alternate digivolution: Lv.7 [Omnimon] for cost 2")
        effect0._alt_digi_cost = 2
        effect0._alt_digi_level = 7
        effect0._alt_digi_name = "Omnimon"

        def condition0(context: Dict[str, Any]) -> bool:
            return True
        effect0.set_can_use_condition(condition0)
        effects.append(effect0)

        # --- Effect 1: [When Digivolving] De-Digivolve 3 one opponent Digimon,
        #     then delete opponent Digimon with total play cost up to 6 ---
        effect1 = ICardEffect()
        effect1.set_timing(EffectTiming.OnEnterFieldAnyone)
        effect1.set_effect_name("EX4-073 De-Digivolve 3 + delete up to cost 6")
        effect1.set_effect_description(
            "[When Digivolving] <De-Digivolve 3> 1 of your opponent's Digimon. "
            "Then, delete up to 6 play cost's total worth of their Digimon."
        )
        effect1.is_when_digivolving = True

        def condition1(context: Dict[str, Any]) -> bool:
            if card and card.permanent_of_this_card() is None:
                return False
            return True
        effect1.set_can_use_condition(condition1)

        def process1(ctx: Dict[str, Any]):
            player = ctx.get('player')
            game = ctx.get('game')
            if not (player and game):
                return
            enemy = player.enemy
            if not enemy:
                return

            # Part 1: De-Digivolve 3 one opponent Digimon
            opp_digimon = [p for p in enemy.battle_area if p.is_digimon]
            if opp_digimon:
                def de_digi_filter(p):
                    return p.is_digimon and p in enemy.battle_area

                def on_de_digi(target_perm):
                    if target_perm and len(target_perm.card_sources) > 1:
                        removed = target_perm.de_digivolve(3)
                        for rc in removed:
                            enemy.trash_cards.append(rc)

                    # Part 2: Delete opponent Digimon with total play cost <= 6
                    _delete_cost_limited(player, enemy, game)

                game.effect_select_opponent_permanent(
                    player, on_de_digi, filter_fn=de_digi_filter,
                    is_optional=False,
                    prompt="Select 1 opponent Digimon to De-Digivolve 3."
                )
            else:
                # No opponent Digimon for de-digivolve, skip to delete
                _delete_cost_limited(player, enemy, game)

        def _delete_cost_limited(player, enemy, game):
            """Delete opponent Digimon with total play cost up to 6.

            Lets the agent select targets one by one. After each selection,
            the remaining budget is checked to constrain further selections.
            """
            budget_remaining = [6]  # mutable container for closure

            def _do_select_loop():
                # Find eligible targets whose cost fits in remaining budget
                eligible = [
                    p for p in enemy.battle_area
                    if p.is_digimon and _get_play_cost(p) <= budget_remaining[0]
                ]
                if not eligible:
                    return

                def cost_filter(p):
                    return (p.is_digimon and p in enemy.battle_area
                            and _get_play_cost(p) <= budget_remaining[0])

                def on_delete(target_perm):
                    cost = _get_play_cost(target_perm)
                    budget_remaining[0] -= cost
                    enemy.delete_permanent(target_perm)
                    # Continue selecting if budget and targets remain
                    _do_select_loop()

                game.effect_select_opponent_permanent(
                    player, on_delete,
                    filter_fn=cost_filter,
                    is_optional=True,
                    prompt=f"Select opponent Digimon to delete (budget: {budget_remaining[0]} remaining)."
                )

            _do_select_loop()

        effect1.set_on_process_callback(process1)
        effects.append(effect1)

        # --- Effect 2: [When Attacking] Trash up to 3 Lv.6+ evo cards,
        #     delete 1 lowest-cost opponent per card trashed,
        #     if 3 trashed then trash top 2 security ---
        effect2 = ICardEffect()
        effect2.set_timing(EffectTiming.OnUseAttack)
        effect2.set_effect_name("EX4-073 Trash evo cards to delete lowest cost + security")
        effect2.set_effect_description(
            "[When Attacking] By trashing up to 3 level 6 or higher cards in "
            "this Digimon's digivolution cards, for each card trashed, "
            "delete 1 of your opponent's Digimon or Tamers with the lowest "
            "play cost. If you trashed 3 cards, trash the top 2 cards of "
            "your opponent's security stack."
        )
        effect2.is_optional = True

        def _is_lv6_plus(c) -> bool:
            level = getattr(c, 'level', None)
            if level is None:
                return False
            has_level = getattr(c, 'has_level', True)
            if not has_level:
                return False
            return level >= 6

        def condition2(context: Dict[str, Any]) -> bool:
            if card and card.permanent_of_this_card() is None:
                return False
            perm = card.permanent_of_this_card()
            ctx_perm = context.get('attacker') or context.get('permanent')
            if perm and ctx_perm and ctx_perm is not perm:
                return False
            if not perm.is_digimon:
                return False
            # Must have at least 1 Lv.6+ evo card
            evo_cards = perm.card_sources[:-1] if len(perm.card_sources) > 1 else []
            has_eligible = any(
                _is_lv6_plus(c) for c in evo_cards
            )
            return has_eligible
        effect2.set_can_use_condition(condition2)

        def process2(ctx: Dict[str, Any]):
            player = ctx.get('player')
            game = ctx.get('game')
            if not (player and game):
                return
            perm = card.permanent_of_this_card() if card else None
            if not perm:
                return
            enemy = player.enemy
            if not enemy:
                return

            from ....data.enums import GamePhase
            from ....game.constants import SOURCES_PER_FIELD

            # Find field index of this permanent
            field_idx = None
            for fi, fp in enumerate(player.battle_area):
                if fp is perm:
                    field_idx = fi
                    break
            if field_idx is None:
                return
            base = 2000 + field_idx * SOURCES_PER_FIELD

            # Let agent select which Lv.6+ evo cards to trash (up to 3)
            evo_cards = perm.card_sources[:-1] if len(perm.card_sources) > 1 else []
            eligible_evo = [c for c in evo_cards if _is_lv6_plus(c)]
            if not eligible_evo:
                return

            trashed = []
            max_trash = min(3, len(eligible_evo))

            # Select evo cards to trash one at a time
            def _select_next_evo():
                if len(trashed) >= max_trash:
                    _do_deletions()
                    return

                current_eligible = [
                    c for c in perm.card_sources[:-1]
                    if _is_lv6_plus(c) and c not in trashed
                ]
                if not current_eligible:
                    _do_deletions()
                    return

                # Build valid action IDs for source selection
                valid = []
                for i, cs in enumerate(perm.card_sources):
                    if cs in current_eligible and (base + i) < 2168:
                        valid.append(base + i)
                if not valid:
                    _do_deletions()
                    return

                def on_select_evo(action_id: int):
                    idx = action_id - base
                    if 0 <= idx < len(perm.card_sources):
                        chosen = perm.card_sources[idx]
                        if chosen in perm.card_sources:
                            perm.card_sources.remove(chosen)
                            player.trash_cards.append(chosen)
                            trashed.append(chosen)
                    _select_next_evo()

                game.request_selection(
                    GamePhase.SelectSource, player, on_select_evo, valid,
                    is_optional=True,
                    prompt=f"Select a Lv.6+ digivolution card to trash ({len(trashed)}/{max_trash})."
                )

            def _do_deletions():
                if not trashed:
                    return

                # For each trashed card, delete 1 opponent Digimon/Tamer
                # with the lowest play cost
                def _delete_lowest_loop(remaining_count):
                    if remaining_count <= 0:
                        # Check security trash
                        if len(trashed) >= 3:
                            for _ in range(min(2, len(enemy.security_cards))):
                                top_sec = enemy.security_cards.pop(0)
                                enemy.trash_cards.append(top_sec)
                        return

                    opp_perms = list(enemy.battle_area)
                    if not opp_perms:
                        if len(trashed) >= 3:
                            for _ in range(min(2, len(enemy.security_cards))):
                                top_sec = enemy.security_cards.pop(0)
                                enemy.trash_cards.append(top_sec)
                        return

                    # Find lowest play cost among opponent permanents
                    def _perm_cost(p):
                        if p.top_card:
                            cost = getattr(p.top_card, 'get_cost_itself', 0)
                            return int(cost) if cost else 99
                        return 99

                    min_cost = min(_perm_cost(p) for p in opp_perms)

                    def lowest_filter(p):
                        return (p.is_digimon or p.is_tamer) and _perm_cost(p) == min_cost

                    def on_delete(target_perm):
                        enemy.delete_permanent(target_perm)
                        _delete_lowest_loop(remaining_count - 1)

                    has_target = any(lowest_filter(p) for p in opp_perms)
                    if has_target:
                        game.effect_select_opponent_permanent(
                            player, on_delete,
                            filter_fn=lowest_filter,
                            is_optional=False)
                    else:
                        _delete_lowest_loop(remaining_count - 1)

                _delete_lowest_loop(len(trashed))

            _select_next_evo()

        effect2.set_on_process_callback(process2)
        effects.append(effect2)

        return effects

File: code/digimon-engine/tests/cards_behavioral/ex4/ex4_073.rs
//! EX4-073 Omnimon Alter-B — Digimon, Lv.7, Black, DP 15000, Cost 15.
//! Traits: Holy Warrior. Attribute: Virus.
//! Evo: Lv6 Black / Cost 5. Special evo: Lv7 w/[Omnimon] in name / Cost 2.
//!
//! # Card text (cards.json)
//!
//! ```text
//! [When Digivolving] <De-Digivolve 3> 1 of your opponent's Digimon.
//!   (Trash up to 3 cards from the top. You can't trash past level 3 cards.)
//!   Then, delete up to 6 play cost's total worth of their Digimon.
//! [When Attacking] By trashing up to 3 level 6 or higher cards in this
//!   Digimon's digivolution cards, for each card trashed, activate the effect
//!   below. Then, if you trashed 3 cards, trash the top 2 cards of your
//!   opponent's security stack.
//!   - Delete 1 of your opponent's Digimon or Tamers with the lowest play cost.
//! ```
//!
//! # DCGO C# reference
//! DCGO/Assets/Scripts/CardEffect/EX4/Black/EX4_073.cs
//!
//! # Patterns this test covers
//! - Two alt-paths (standard digivolve + Omnimon-name special digivolve at cost 2)
//! - [When Digivolving] arm 1: select_opponent_permanent + de_digivolve(3, stop_at:3)
//! - [When Digivolving] arm 2: BLOCKED — G-MULTI-SELECT-OPP-PLAY-COST-SUM
//!   (no play_cost-budget multi-select in DSL or engine; only DP-budget exists)
//! - [When Attacking] entire clause: BLOCKED — G-PLAY-COST-AGGREGATE
//!   (no `lowest_play_cost` predicate in DSL; sibling of literal P-189
//!   G-PLAY-COST-LTE). G-DSL-SELECT-OWN-SOURCES-FILTER was closed 2026-05-08
//!   (select_own_sources now accepts filter: + evaluates against source cards),
//!   so the outer source-trash step is no longer blocked; only the inner
//!   per-trash delete (lowest_play_cost aggregate) and the count-binding
//!   for the conditional security-trash remain blocked on G-PLAY-COST-AGGREGATE.

use digimon_dsl::compiled::{
    CompiledClause, CompiledScope, CompiledTiming, CompiledTriggeredClause,
};
use digimon_engine::debug_runner::{make_test_card, DebugRunner};
use digimon_engine::enums::EffectTiming;
use digimon_engine::selection::{SelectionKind, TriggerSource};

// ─── Helper ──────────────────────────────────────────────────────────────────

fn runner() -> DebugRunner {
    DebugRunner::builder()
        .dsl_card("EX4-073")
        .expect("EX4-073 in embedded DSL pack")
        .memory(20)
        .start()
}

// ═══════════════════════════════════════════════════════════════════════════════
// Section 1 — Compilation smoke test
// ═══════════════════════════════════════════════════════════════════════════════

/// EX4-073 YAML must compile and appear in the embedded DSL pack.
#[test]
fn ex4_073_yaml_compiles_without_error() {
    let runner = runner();
    let card = runner.compiled_card("EX4-073");
    assert!(
        card.is_some(),
        "EX4-073 must be present in the embedded DSL pack (YAML must parse + compile)"
    );
}

// ═══════════════════════════════════════════════════════════════════════════════
// Section 2 — Structural assertions
// ═══════════════════════════════════════════════════════════════════════════════

/// EX4-073 has exactly one triggered clause authored today (clause B partial —
/// arm 1 only). Clause C ([When Attacking]) is OMITTED entirely under blocking
/// gaps; arm 2 of clause B (delete-up-to-6-cost) is OMITTED under
/// G-MULTI-SELECT-OPP-PLAY-COST-SUM.
///
/// When G-MULTI-SELECT-OPP-PLAY-COST-SUM closes, arm 2 is appended to the
/// existing when_digivolving clause (still 1 triggered clause).
/// When the [When Attacking] gaps close, a second triggered clause is added
/// (then 2 triggered clauses).
#[test]
fn ex4_073_has_one_triggered_clause_authored() {
    let runner = runner();
    let card = runner
        .compiled_card("EX4-073")
        .expect("EX4-073 in embedded pack");

    let triggered: Vec<&CompiledTriggeredClause> = card
        .effects
        .iter()
        .filter_map(|c| match c {
            CompiledClause::Triggered(t) => Some(t),
            _ => None,
        })
        .collect();
    assert_eq!(
        triggered.len(),
        1,
        "expected 1 triggered clause authored today (clause B arm 1 only); \
         clause B arm 2 + clause C are BLOCKED — see YAML header comments"
    );
}

/// The single authored triggered clause fires on WhenDigivolving and is
/// mandatory (no "you may"), face-up scope, no [Once Per Turn].
#[test]
fn ex4_073_clause_b_fires_on_when_digivolving_mandatory_face_up() {
    let runner = runner();
    let card = runner
        .compiled_card("EX4-073")
        .expect("EX4-073 in embedded pack");

    let clause = card
        .effects
        .iter()
        .filter_map(|c| match c {
            CompiledClause::Triggered(t) => Some(t),
            _ => None,
        })
        .find(|t| t.when.contains(&CompiledTiming::WhenDigivolving))
        .expect("clause B must include WhenDigivolving");

    assert!(
        !clause.optional,
        "clause B is mandatory — printed text has no 'you may' on the outer trigger"
    );
    assert!(!clause.once_per_turn, "clause B has no [Once Per Turn]");
    assert_eq!(clause.scope, CompiledScope::FaceUp, "clause B is own-scope");
}

/// EX4-073 has two alt-paths: standard digivolve (Lv6 Black / 5) and
/// Omnimon-name special digivolve (Lv7 / 2).
#[test]
fn ex4_073_has_two_alt_paths() {
    let runner = runner();
    let card = runner
        .compiled_card("EX4-073")
        .expect("EX4-073 in embedded pack");

    assert_eq!(
        card.alt_paths.len(),
        2,
        "EX4-073 has standard digivolve (Lv6 Black) + Omnimon-name special (Lv7) = 2 alt-paths"
    );
}

// ═══════════════════════════════════════════════════════════════════════════════
// Section 3 — Clause B arm 1 behavioral: <De-Digivolve 3> on 1 opp Digimon
// ═══════════════════════════════════════════════════════════════════════════════

/// Clause B arm 1: opponent has 1 Lv5 Digimon with stacked sources →
/// fire WhenDigivolving → selection is installed for the de-digivolve target.
/// (No multi-pick budget — clause B arm 2 is BLOCKED.)
#[test]
fn ex4_073_clause_b_installs_selection_for_de_digivolve_target() {
    let mut opp_lv5 = make_test_card("OPP-LV5", "OppLv5");
    opp_lv5.level = Some(5);
    opp_lv5.dp = Some(7000);

    let mut runner = DebugRunner::builder()
        .dsl_card("EX4-073")
        .expect("EX4-073 in embedded DSL pack")
        .add_card(opp_lv5)
        .memory(20)
        .start();

    let handle = runner.place_on_field(0, "EX4-073", Some(0));
    runner.place_on_field(1, "OPP-LV5", Some(0));

    runner.game.enqueue_triggered(
        EffectTiming::WhenDigivolving,
        TriggerSource::Permanent(handle),
    );
    runner.game.drain_effect_queue();

    // A pending selection must be installed for the de-digivolve pick.
    assert!(
        runner.game.pending_selection.is_some(),
        "clause B arm 1 must install a pending selection for the de-digivolve target"
    );
}

/// Clause B arm 1: opponent has no Digimon → outer condition gate prevents
/// firing; no selection installed.
#[test]
fn ex4_073_clause_b_no_fire_when_opponent_has_no_digimon() {
    let mut runner = runner();
    let handle = runner.place_on_field(0, "EX4-073", Some(0));

    runner.game.enqueue_triggered(
        EffectTiming::WhenDigivolving,
        TriggerSource::Permanent(handle),
    );
    runner.game.drain_effect_queue();

    assert!(
        runner.game.pending_selection.is_none(),
        "no opponent Digimon → outer existential condition fails → no selection"
    );
}

/// Clause B arm 1: the installed selection must be `SelectionKind::OppField`
/// and mandatory (`is_optional == false`).
///
/// DCGO: `canNoSelect: false` — printed text has no "you may" on the outer
/// select; if a legal target exists the player must pick one.
#[test]
fn ex4_073_clause_b_selection_is_oppfield_and_mandatory() {
    let mut opp_lv5 = make_test_card("OPP-LV5B", "OppLv5B");
    opp_lv5.level = Some(5);
    opp_lv5.dp = Some(7000);

    let mut runner = DebugRunner::builder()
        .dsl_card("EX4-073")
        .expect("EX4-073 in embedded DSL pack")
        .add_card(opp_lv5)
        .memory(20)
        .start();

    let handle = runner.place_on_field(0, "EX4-073", Some(0));
    runner.place_on_field(1, "OPP-LV5B", Some(0));

    runner.game.enqueue_triggered(
        EffectTiming::WhenDigivolving,
        TriggerSource::Permanent(handle),
    );
    runner.game.drain_effect_queue();

    let kind = runner
        .pending_kind()
        .expect("a pending selection must be installed when opp has a Digimon");
    assert_eq!(
        kind,
        SelectionKind::OppField,
        "select_opponent_permanent lowers to SelectionKind::OppField"
    );
    assert!(
        !runner.pending_is_optional(),
        "De-Digivolve target selection must be mandatory (DCGO canNoSelect: false)"
    );
}

/// Clause B arm 1 behavioral outcome: resolving the de-digivolve target
/// trashes up to 3 cards from the top of the selected Digimon's stack.
///
/// Setup: opponent has a Lv6 Digimon with 3 sources stacked below it
/// (total 4 cards in stack). After de_digivolve(3, stop_at:3), the
/// stack should have been reduced by 3 (to 1 base card, since all
/// intermediate sources are above the Lv3 floor).
///
/// The Lv3 base card itself is never trashed (stop_at_level: 3 prevents
/// popping when the NEXT top would be below Lv3).
#[test]
fn ex4_073_clause_b_de_digivolve_reduces_opp_stack_by_up_to_3() {
    // Opponent: a Lv6 top card stacked on Lv5 / Lv4 / Lv3 sources.
    // Stack (bottom→top): Lv3_base | Lv4_src | Lv5_src | Lv6_top
    // De-Digivolve 3 with stop_at:3:
    //   Pop Lv6 → next top would be Lv5 (≥3) → OK.
    //   Pop Lv5 → next top would be Lv4 (≥3) → OK.
    //   Pop Lv4 → next top would be Lv3 (≥3) → OK.
    //   amount cap reached (3 pops) → done.
    // Remaining stack: Lv3_base (1 card).
    let mut opp_top = make_test_card("OPP-LV6-TOP", "OppLv6Top");
    opp_top.level = Some(6);
    opp_top.dp = Some(9000);

    let mut src_lv5 = make_test_card("OPP-SRC-LV5", "OppSrcLv5");
    src_lv5.level = Some(5);

    let mut src_lv4 = make_test_card("OPP-SRC-LV4", "OppSrcLv4");
    src_lv4.level = Some(4);

    let mut src_lv3 = make_test_card("OPP-SRC-LV3", "OppSrcLv3");
    src_lv3.level = Some(3);

    let mut runner = DebugRunner::builder()
        .dsl_card("EX4-073")
        .expect("EX4-073 in embedded DSL pack")
        .add_card(opp_top)
        .add_card(src_lv5.clone())
        .add_card(src_lv4.clone())
        .add_card(src_lv3.clone())
        .memory(20)
        .start();

    let owner_handle = runner.place_on_field(0, "EX4-073", Some(0));
    let opp_perm = runner.place_on_field(1, "OPP-LV6-TOP", Some(0));

    // Stack 3 sources below the top card (bottom→top order: lv3 / lv4 / lv5).
    runner.push_source(opp_perm, "OPP-SRC-LV3");
    runner.push_source(opp_perm, "OPP-SRC-LV4");
    runner.push_source(opp_perm, "OPP-SRC-LV5");

    // Verify setup: 4 cards in stack before the trigger.
    let stack_before = runner.game.player(1).battle_area[opp_perm.index as usize]
        .card_sources
        .len();
    assert_eq!(stack_before, 4, "setup: opp permanent must have 4 cards (3 sources + top)");

    runner.game.enqueue_triggered(
        EffectTiming::WhenDigivolving,
        TriggerSource::Permanent(owner_handle),
    );
    runner.game.drain_effect_queue();

    // Resolve the OppField de-digivolve target pick.
    let view = runner
        .pending_selection_view()
        .expect("OppField selection must be installed");
    let sel_player = view.selecting_player;
    let action_id = view.valid_action_ids[0];
    runner
        .game
        .resolve_selection(sel_player, action_id)
        .expect("target pick resolves cleanly");

    // No further selections after de_digivolve (clause B arm 2 is BLOCKED).
    assert!(
        runner.game.pending_selection.is_none(),
        "after de_digivolve resolves, no further selections should remain \
         (clause B arm 2 is BLOCKED on G-MULTI-SELECT-OPP-PLAY-COST-SUM)"
    );

    // Stack must now be reduced by 3 (3 sources popped; Lv3 base remains).
    let stack_after = runner.game.player(1).battle_area[opp_perm.index as usize]
        .card_sources
        .len();
    assert_eq!(
        stack_after,
        stack_before - 3,
        "de_digivolve(3, stop_at:3) must trash exactly 3 sources from the top \
         (Lv6, Lv5, Lv4 popped; Lv3 base remains)"
    );
}

/// Clause B arm 1 stop-at-Lv3: when the next card below the top is Lv2
/// (i.e. below the floor), de_digivolve stops after 0 pops even though
/// the amount cap allows up to 3.
///
/// Setup: opponent has a Lv4 Digimon with a Lv2 source below it
/// (total 2 cards). De-Digivolve 3 with stop_at:3: popping the Lv4
/// would leave the Lv2 as top — which is < 3, so it immediately stops.
/// Stack remains at 2 cards (unchanged).
///
/// Note: the target selection still fires (the outer existential
/// condition only checks that an opp Digimon exists); it's the
/// de_digivolve step itself that evaluates the floor.
#[test]
fn ex4_073_clause_b_de_digivolve_stops_at_lv3_floor_when_next_top_below_3() {
    let mut opp_lv4 = make_test_card("OPP-LV4-STOP", "OppLv4Stop");
    opp_lv4.level = Some(4);
    opp_lv4.dp = Some(6000);

    let mut src_lv2 = make_test_card("OPP-SRC-LV2", "OppSrcLv2");
    src_lv2.level = Some(2);

    let mut runner = DebugRunner::builder()
        .dsl_card("EX4-073")
        .expect("EX4-073 in embedded DSL pack")
        .add_card(opp_lv4)
        .add_card(src_lv2)
        .memory(20)
        .start();

    let owner_handle = runner.place_on_field(0, "EX4-073", Some(0));
    let opp_perm = runner.place_on_field(1, "OPP-LV4-STOP", Some(0));
    runner.push_source(opp_perm, "OPP-SRC-LV2");

    // Verify setup: 2 cards (Lv2 source at bottom, Lv4 top).
    let stack_before = runner.game.player(1).battle_area[opp_perm.index as usize]
        .card_sources
        .len();
    assert_eq!(stack_before, 2, "setup: opp permanent must have 2 cards");

    runner.game.enqueue_triggered(
        EffectTiming::WhenDigivolving,
        TriggerSource::Permanent(owner_handle),
    );
    runner.game.drain_effect_queue();

    // Resolve the OppField target pick.
    let view = runner
        .pending_selection_view()
        .expect("OppField selection must install — the existential condition passes");
    let sel_player = view.selecting_player;
    let action_id = view.valid_action_ids[0];
    runner
        .game
        .resolve_selection(sel_player, action_id)
        .expect("target pick resolves");

    // de_digivolve must have popped 0 cards: next top would be Lv2 < 3.
    let stack_after = runner.game.player(1).battle_area[opp_perm.index as usize]
        .card_sources
        .len();
    assert_eq!(
        stack_after, stack_before,
        "stop_at_level:3 prevents trashing when next top would be Lv2 (< 3); \
         stack must remain unchanged"
    );
}

// ═══════════════════════════════════════════════════════════════════════════════
// Section 4 — Clause B arm 2 (BLOCKED): delete up to 6 play-cost worth
// ═══════════════════════════════════════════════════════════════════════════════

/// Clause B arm 2 — "delete up to 6 play cost's total worth of their Digimon".
///
/// BLOCKED on G-MULTI-SELECT-OPP-PLAY-COST-SUM: the engine has only the
/// DP-budget multi-select primitive (`select_opponent_permanents_by_dp_budget`)
/// and the matching DSL step `select_opponent_dp_budget`. There is no
/// play-cost-budget sibling. Authored as #[ignore] until the gap closes;
/// when it does, the test exercises a multi-select that:
///   - lets the player pick 0..N opp Digimon,
///   - enforces running play_cost sum ≤ 6 after each pick,
///   - filters out any single Digimon whose individual play_cost > 6,
///   - deletes all picked permanents on resolution.
#[test]
#[ignore = "pending: G-MULTI-SELECT-OPP-PLAY-COST-SUM — no play-cost-budget multi-select API in engine or DSL (only DpBudget exists)"]
fn ex4_073_clause_b_arm2_deletes_opp_digimon_within_play_cost_sum_6() {
    todo!("write once G-MULTI-SELECT-OPP-PLAY-COST-SUM is closed");
}

/// Clause B arm 2: a single opp Digimon with play_cost > 6 must be ineligible
/// (individual cap, mirrors DCGO `permanent.TopCard.GetCostItself <= 6`).
/// BLOCKED — same gap.
#[test]
#[ignore = "pending: G-MULTI-SELECT-OPP-PLAY-COST-SUM — per-card cap requires the play-cost-budget primitive's candidate filter"]
fn ex4_073_clause_b_arm2_excludes_opp_digimon_with_play_cost_above_6() {
    todo!("write once G-MULTI-SELECT-OPP-PLAY-COST-SUM is closed");
}

/// Clause B arm 2: when no opp Digimon are eligible (all cost > 6 or none on
/// field), the arm silently no-ops and arm 1 still resolved. BLOCKED — same gap.
#[test]
#[ignore = "pending: G-MULTI-SELECT-OPP-PLAY-COST-SUM — silent no-op behavior depends on the primitive's existential gate"]
fn ex4_073_clause_b_arm2_silent_noop_when_no_eligible_targets() {
    todo!("write once G-MULTI-SELECT-OPP-PLAY-COST-SUM is closed");
}

// ═══════════════════════════════════════════════════════════════════════════════
// Section 5 — Clause C (BLOCKED): [When Attacking] trash sources + lowest-cost
// ═══════════════════════════════════════════════════════════════════════════════

/// Clause C outer step — "trash up to 3 level 6 or higher cards in this
/// Digimon's digivolution cards".
///
/// G-DSL-SELECT-OWN-SOURCES-FILTER is CLOSED (2026-05-08): `select_own_sources`
/// now accepts `filter:` and evaluates it against each source card via
/// `eval_predicate_with_bindings(PredicateSubject::Card(source.card))`.
/// The `level_gte: 6` filter is now representable.
///
/// The clause body remains BLOCKED on G-PLAY-COST-AGGREGATE: the inner
/// per-trash delete ("delete 1 opp Digimon/Tamer with the lowest play cost")
/// requires a `lowest_play_cost` aggregate predicate that does not yet exist
/// in the DSL (only literal `play_cost_lte` is available). Until the inner
/// body is writable, the clause cannot be authored faithfully.
#[test]
#[ignore = "pending: G-PLAY-COST-AGGREGATE — inner per-trash delete requires `lowest_play_cost` aggregate predicate (no `LowestPlayCost` in AggregateSelector; DSL has only literal play_cost_lte). Outer source-trash step is unblocked (G-DSL-SELECT-OWN-SOURCES-FILTER closed 2026-05-08)."]
fn ex4_073_clause_c_trashes_up_to_3_lv6_or_higher_sources() {
    todo!("write once G-PLAY-COST-AGGREGATE is closed");
}

/// Clause C inner per-trash effect — "Delete 1 of your opponent's Digimon or
/// Tamers with the lowest play cost." BLOCKED on G-PLAY-COST-AGGREGATE: the
/// DSL has the LITERAL `play_cost_lte` form (closed by P-189 on 2026-05-01)
/// but no aggregate `lowest_play_cost: bool` predicate analogous to
/// `lowest_dp` / `highest_dp` / `lowest_level`. BT9-112 hit the same gap and
/// uses a per-card raw_rust escape hatch; new authorings should not extend
/// that pattern.
#[test]
#[ignore = "pending: G-PLAY-COST-AGGREGATE — no `lowest_play_cost` predicate (sibling of `lowest_dp`); BT9-112 raw_rust hatch is grandfathered, not extended"]
fn ex4_073_clause_c_inner_deletes_opp_lowest_play_cost_per_trash() {
    todo!("write once G-PLAY-COST-AGGREGATE is closed");
}

/// Clause C tail — "if you trashed 3 cards, trash the top 2 cards of your
/// opponent's security stack" — transitively BLOCKED on G-PLAY-COST-AGGREGATE:
/// the count binding comes from the inner for-each / source-trash loop, which
/// cannot be authored until the inner delete (lowest_play_cost) is closed.
/// G-DSL-SELECT-OWN-SOURCES-FILTER is CLOSED (2026-05-08) so the outer
/// source-trash step itself is no longer the bottleneck.
#[test]
#[ignore = "pending: G-PLAY-COST-AGGREGATE (parent gap) — count binding for the 3-trash threshold depends on the for-each source-trash body, which requires the `lowest_play_cost` aggregate predicate to be representable"]
fn ex4_073_clause_c_trashes_2_opp_security_when_3_sources_trashed() {
    todo!("write once G-PLAY-COST-AGGREGATE is closed");
}

/// Clause C is OPTIONAL ("By trashing ... up to 3" — DCGO isOptional: true,
/// canNoSelect: () => true on the source-select). When the gaps close, the
/// player should be able to pass on the entire clause.
///
/// G-DSL-SELECT-OWN-SOURCES-FILTER is CLOSED (2026-05-08). The clause remains
/// BLOCKED solely on G-PLAY-COST-AGGREGATE (inner delete body).
#[test]
#[ignore = "pending: G-PLAY-COST-AGGREGATE — clause body requires `lowest_play_cost` predicate; clause optionality test is writable once the body is writable. G-DSL-SELECT-OWN-SOURCES-FILTER closed 2026-05-08."]
fn ex4_073_clause_c_player_may_decline_to_trash_any_sources() {
    todo!("write once G-PLAY-COST-AGGREGATE is closed");
}
